# assignment 1


In [1]:
import numpy as np
import pandas as pd 
import matplotlib as mtp 
import seaborn as sns 
import re
import string
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')


[nltk_data] Downloading package stopwords to C:\Users\Saloni
[nltk_data]     Kumari\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

## extract the refrence dataset from kaggle

In [2]:
df=pd.read_csv("C:/Users/Saloni Kumari/.vscode/SMILES_Big_Data_Set.csv")

In [3]:
df.head()

SMILES  pIC50  \
0         O=S(=O)(Nc1cccc(-c2cnc3ccccc3n2)c1)c1cccs1   4.26   
1  O=c1cc(-c2nc(-c3ccc(-c4cn(CCP(=O)(O)O)nn4)cc3)...   4.34   
2             NC(=O)c1ccc2c(c1)nc(C1CCC(O)CC1)n2CCCO   4.53   
3                NCCCn1c(C2CCNCC2)nc2cc(C(N)=O)ccc21   4.56   
4                  CNC(=S)Nc1cccc(-c2cnc3ccccc3n2)c1   4.59   

                                                mol  num_atoms    logP  
0  <rdkit.Chem.rdchem.Mol object at 0x7f59df45bc30>         25  4.1591  
1  <rdkit.Chem.rdchem.Mol object at 0x7f59a320c9e0>         36  3.6743  
2  <rdkit.Chem.rdchem.Mol object at 0x7f59a320cac0>         23  1.5361  
3  <rdkit.Chem.rdchem.Mol object at 0x7f59a320cba0>         22  0.9510  
4  <rdkit.Chem.rdchem.Mol object at 0x7f59a320c7b0>         21  3.2130

In [4]:
df.tail()

SMILES  pIC50  \
16082   S=C(NN=C(c1ccccn1)c1ccccn1)Nc1ccccc1   0.00   
16083                S=C=NCCCCCCCCCCc1ccccc1   0.00   
16084                  S=C=NCCCCCCCCc1ccccc1   0.00   
16085  S=c1[nH]nc(Cn2ccc3ccccc32)n1-c1ccccc1   0.00   
16086       N=[N+]=NC=Cc1ccc([N+](=O)[O-])o1   0.02   

                                                    mol  num_atoms     logP  
16082  <rdkit.Chem.rdchem.Mol object at 0x7f59a314ed50>         24  3.21560  
16083  <rdkit.Chem.rdchem.Mol object at 0x7f59a314edc0>         19  5.45270  
16084  <rdkit.Chem.rdchem.Mol object at 0x7f59a314ee30>         17  4.67250  
16085  <rdkit.Chem.rdchem.Mol object at 0x7f59a314eea0>         22  3.93289  
16086  <rdkit.Chem.rdchem.Mol object at 0x7f59a314ef10>         13  1.70887

In [5]:
df.duplicated().sum()

np.int64(0)

## so, there is no duplicate data 


In [6]:
df.isnull().sum()

SMILES          0
pIC50        1050
mol             0
num_atoms       0
logP            0
dtype: int64

1.Only the column pIC50 has 1050 missing (null) values.

2.All other columns are fully populated (no missing data).
    Now we have to drop all this

In [7]:
df=df.dropna()

In [8]:
df['SMILES'].unique()
df['SMILES'].value_counts()

SMILES
CC(NC1CCC(n2cc(C(N)=O)c(Nc3ccc(S(=O)(=O)C(F)(F)F)cc3)n2)C(C#N)C1)C1CC1    6
N#CC1CC(N2CCC2)CCC1n1cc(C(N)=O)c(Nc2ccc(C(O)C(F)(F)F)cc2)n1               4
N#CC1CC(N2CC3(CCC3)C2)CCC1n1cc(C(N)=O)c(Nc2ccc(C(O)C(F)(F)F)cc2)n1        4
N#CC1CC(OCC2CC2)CCC1n1cc(C(N)=O)c(Nc2ccc(C(O)C(F)(F)F)cc2)n1              4
N#CC1CC(NC(C2CC2)C(F)(F)F)CCC1n1cc(C(N)=O)c(Nc2ccc(C(F)(F)F)cc2)n1        4
                                                                         ..
O=c1cc(CSc2ccccn2)occ1O                                                   1
O=C(c1c(O)c2ccc(Cl)cc2[nH]c1=O)c1csc2ccccc12                              1
O=c1[nH]c2cc(Cl)ccc2c(O)c1-c1cccc(NCc2ccccc2)c1                           1
O=C(Cc1ccccc1)OC1Cc2c(O)cc(O)cc2OC1c1ccc(O)c(O)c1                         1
N=[N+]=NC=Cc1ccc([N+](=O)[O-])o1                                          1
Name: count, Length: 14823, dtype: int64

In [9]:
inconsistent = df.groupby('SMILES')['pIC50'].nunique()
inconsistent = inconsistent[inconsistent > 1]
df[df['SMILES'].isin(inconsistent.index)]

SMILES  pIC50  \
67    O=c1[nH]c2cnc(-n3cnc4ccccc43)nc2n1C1CCOc2c(F)c...   5.16   
162   CC(=O)NC1CCc2ccc(Oc3cnc4[nH]cc(C(=O)NC(C)C)c4n...   5.77   
172   O=C1NC(=O)C(c2c[nH]c3ccccc23)=C1Nc1cccc(OC2CCN...   5.83   
192   O=C1NC(=O)C(c2c[nH]c3ccccc23)=C1Nc1cccc(OC2CCN...   5.91   
211   CN(C)C1CCC(n2cc(C(N)=O)c(Nc3ccc(OC(F)F)nc3)n2)...   5.96   
...                                                 ...    ...   
1889          CC(=O)N1CCC(Nc2ncccc2-c2cnc3[nH]ccc3n2)C1  10.40   
1892  CN(c1[nH]cnc2nccc1-2)C1CCC(CS(=O)(=O)N2CCC(C)(...  10.44   
1897      CS(=O)(=O)N1CCC(Nc2ncccc2-c2cnc3[nH]ccc3n2)C1  10.53   
1899         COC(=O)N1CCC(Nc2ncccc2-c2cnc3[nH]ccc3n2)C1  10.56   
1909      CS(=O)(=O)N1CCC(Nc2ncccc2-c2cnc3[nH]ccc3n2)C1  10.97   

                                                   mol  num_atoms     logP  
67    <rdkit.Chem.rdchem.Mol object at 0x7f59a320e650>         30  2.96950  
162   <rdkit.Chem.rdchem.Mol object at 0x7f59a31dfa70>         29  3.01180  
172   <rdkit.Chem.rdchem.Mol object at 0x7f59a31dfed0>         36  3.83930  
192   <rdkit.Chem.rdchem.Mol object at 0x7f59a2938820>         36  3.83930  
211   <rdkit.Chem.rdchem.Mol object at 0x7f59a2939070>         30  2.51698  
...                                                ...        ...      ...  
1889  <rdkit.Chem.rdchem.Mol object at 0x7f59a29674c0>         24  2.05260  
1892  <rdkit.Chem.rdchem.Mol object at 0x7f59a2967610>         28  1.69100  
1897  <rdkit.Chem.rdchem.Mol object at 0x7f59a2967840>         25  1.46570  
1899  <rdkit.Chem.rdchem.Mol object at 0x7f59a2967920>         25  2.27250  
1909  <rdkit.Chem.rdchem.Mol object at 0x7f59a2967d80>         25  1.46570  

[340 rows x 5 columns]

# There are some inconsistent data and we have to remove that
  because it Violates the One-to-One Mapping

In [10]:
df = df[~df['SMILES'].isin(inconsistent.index)]
print("Shape after removing inconsistencies:", df.shape)
print("Remaining duplicates in SMILES:", df['SMILES'].duplicated().sum())


Shape after removing inconsistencies: (14697, 5)
Remaining duplicates in SMILES: 27


# yes, there is some duplicate data
    These don’t add new information. They just repeat the same data point.
    So, we Should be removed to avoid bias or redundancy.



In [11]:
df = df.drop_duplicates(subset=['SMILES', 'pIC50'])

In [12]:
df['SMILES'] = df['SMILES'].str.strip().str.lower()


# Now ,I want to print the cleaned version


In [13]:
print("Shape:", df.shape)
print("Missing values:\n", df.isnull().sum())
df = df.drop_duplicates(subset=['SMILES', 'pIC50'])
print("Duplicate SMILES:", df['SMILES'].duplicated().sum())

Shape: (14670, 5)
Missing values:
 SMILES       0
pIC50        0
mol          0
num_atoms    0
logP         0
dtype: int64
Duplicate SMILES: 32


### That means there are still some duplicate SMILES, but with different pIC50 values, which drop_duplicates() didn’t remove — because it only drops exact duplicate pairs.

In [14]:
df.head()

SMILES  pIC50  \
0         o=s(=o)(nc1cccc(-c2cnc3ccccc3n2)c1)c1cccs1   4.26   
1  o=c1cc(-c2nc(-c3ccc(-c4cn(ccp(=o)(o)o)nn4)cc3)...   4.34   
2             nc(=o)c1ccc2c(c1)nc(c1ccc(o)cc1)n2ccco   4.53   
3                ncccn1c(c2ccncc2)nc2cc(c(n)=o)ccc21   4.56   
4                  cnc(=s)nc1cccc(-c2cnc3ccccc3n2)c1   4.59   

                                                mol  num_atoms    logP  
0  <rdkit.Chem.rdchem.Mol object at 0x7f59df45bc30>         25  4.1591  
1  <rdkit.Chem.rdchem.Mol object at 0x7f59a320c9e0>         36  3.6743  
2  <rdkit.Chem.rdchem.Mol object at 0x7f59a320cac0>         23  1.5361  
3  <rdkit.Chem.rdchem.Mol object at 0x7f59a320cba0>         22  0.9510  
4  <rdkit.Chem.rdchem.Mol object at 0x7f59a320c7b0>         21  3.2130

In [15]:
df['SMILES'] = df['SMILES'].apply(lambda x: re.sub(r"http\S+|www\S+|@\w+|#\w+", '', x))
df['SMILES'] = df['SMILES'].apply(lambda x: re.sub(r"[^A-Za-z0-9\s]", '', x))

### Now,URLS,hashtags,Emojis,and special characters are removed

In [16]:
df['SMILES'] = df['SMILES'].apply(lambda x: x.translate(str.maketrans('', '', string.punctuation)))
df['SMILES'] = df['SMILES'].str.strip()

### This help to remove punctuation and spaces

In [17]:
df['SMILES'] = df['SMILES'].str.lower()

In [18]:
stop_words = set(stopwords.words('english'))
df['SMILES'] = df['SMILES'].apply(lambda x: ' '.join([word for word in x.split() if word not in stop_words]))

### stop words are removed by this

In [19]:
df.head()

SMILES  pIC50  \
0                 osonc1ccccc2cnc3ccccc3n2c1c1cccs1   4.26   
1  oc1ccc2ncc3cccc4cnccpooonn4cc3nhc2c2cccfcc2ccnh1   4.34   
2                     ncoc1ccc2cc1ncc1cccocc1n2ccco   4.53   
3                      ncccn1cc2ccncc2nc2cccnoccc21   4.56   
4                       cncsnc1ccccc2cnc3ccccc3n2c1   4.59   

                                                mol  num_atoms    logP  
0  <rdkit.Chem.rdchem.Mol object at 0x7f59df45bc30>         25  4.1591  
1  <rdkit.Chem.rdchem.Mol object at 0x7f59a320c9e0>         36  3.6743  
2  <rdkit.Chem.rdchem.Mol object at 0x7f59a320cac0>         23  1.5361  
3  <rdkit.Chem.rdchem.Mol object at 0x7f59a320cba0>         22  0.9510  
4  <rdkit.Chem.rdchem.Mol object at 0x7f59a320c7b0>         21  3.2130